### PR TITLE
Validation fixes

### LINKED ISSUE #41: Which validation metrics to use
One of the bigger issues we face is the impact of low-probability, high-SNR glitches increasing our false alarm rate. While we want to be monitoring our performance on the "true" data distribution during training, it also might be helpful to evaluate our efficacy on these events which don't occur frequently enough to allow a meaningful sample size from the lengths of background data we can reasonably do validation on.

One solution might be to create a dataset of _only_ high (for some definition of "high") SNR glitches that gets iterated through sequentially during a separate validation step. For the data-loading portion, it might be as simple as rejiggering the `GlitchSampler` to be able to sample deterministically and sequentially.
- What does "sequentially" mean here? Do we sample each kernel (or sample with some stride) that contains the glitch "trigger"? Or just produce one kernel per glitch?
- How does the answer to that question impact how we evaluate our performance? Kernels from the same glitch aren't independent, do we instead do some sort of accumulation across these kernels to mimic the matched filtering inference approach? And compare against what, a similar analysis of the validation injections?

### LINKED ISSUE #170: Improve validation during training

### PR DESCRIPTION
**Pull request by alecgunny**
Implement measurements of recall (fraction of signals recovered) vs. background and glitch datasets separately. Closes #41 , closes #170 . Re-creation of #221 but with the git history better aligned.

For background, outputs are pooled across adjustable windows at an adjustable stride to simulate test-time clustering, and recall is measured using thresholds from the top `k` background "events".

For glitches, recall is measured using thresholds from adjustable specificity levels (fraction of glitches correctly classified at a given threshold).

Monitoring (for early stopping and model selection) can be performed against any threshold for either metric. Checkpointing at a specified cadence has been added as well.

- [x] Add `Recorder` and `Validator` parameters to `train:main` function definition
- [x] Add tests
- [x] Fix `sample_rate` used for background pooling to be `1 / stride`

### CODE REVIEW

**Comment by alecgunny** [comment]
Tests aren't exhaustive, but cover the most relevant functionality.

File: projects/sandbox/train/tests/test_validation.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    # after pooling, the top 3 values should come
+    # out to 2, 1.5, and 0.8
+    background[3] = 1
+    background[4] = 1.5
+    background[10] = 2
+    background[14] = 0.5
+    background[18] = 0.9
+    background[22] = 0.8
+    background[30] = 3  # this will get trimmed
```
Comment (EthanMarx) [inline projects/sandbox/train/tests/test_validation.py:18]: is this a `torch.nn.functional.maxpool1d` artifact?

**Review by EthanMarx — APPROVED** [review]

File: projects/sandbox/train/train/train.py [hunk 2]
```diff
@@ -204,9 +234,56 @@ def main(
     # into one file for simplicity
     background = load_background(hanford_background, livingston_background)
     if valid_frac is not None:
-        background, valid_background = split(background, 1 - valid_frac, 1)
-        valid_injector.fit(H1=background[0], L1=background[1])
-        valid_waveforms, _ = valid_injector.sample(-1)
+        # split up our background data into train and validation splits
+        background, valid_background = split(background, 1 - valid_frac, -1)
```
[the remark this answers is not in the document]
Comment (alecgunny) [reply projects/sandbox/train/train/train.py:238]: Yeah in this case the background array is 2D so those should be equivalent. I don't know exactly why I opted for this change, but now that I'm thinking about it I do like the idea of `-1` as a handy shorthand in our context for "do this operation along the time dimension," since that's where it will be regardless of the shape of the array

File: projects/sandbox/train/tests/test_validation.py (hunk 1 above)
Comment (alecgunny) [reply projects/sandbox/train/tests/test_validation.py:18]: Yeah exactly, actually didn't know this until I built this but I guess it adopts the convention of sloughing data off the end that can't be fit into a full kernel rather than doing the op on a shorter kernel. Good to keep in mind for the future I guess
